File: jserver-core/src/main/java/com/github/gr3gdev/jserver/http/impl/ReaderUtil.java
package com.github.gr3gdev.jserver.http.impl;

import com.github.gr3gdev.jserver.http.Request;

import java.io.BufferedReader;
import java.io.IOException;
import java.util.Optional;
import java.util.StringTokenizer;

/**
 * ReaderUtil.
 *
 * @author Gregory Tardivel
 */
public class ReaderUtil {

    public static void loadHeaders(Request request, BufferedReader pReader) throws IOException {
        String headerLine = pReader.readLine();
        while (headerLine != null && !headerLine.isBlank()) {
            final StringTokenizer hTokens = new StringTokenizer(headerLine, ":");
            if (hTokens.hasMoreTokens()) {
                final String key = hTokens.nextToken();
                if (hTokens.hasMoreTokens()) {
                    request.headers(key, hTokens.nextToken().trim());
                }
            }
            headerLine = pReader.readLine();
        }
    }

    public static void loadParameters(Request request, String pathParameters, BufferedReader pReader) throws IOException {
        final Optional<String> contentType = request.headers("Content-Type");
        final StringBuilder payload = new StringBuilder();
        while (pReader.ready()) {
            payload.append((char) pReader.read());
        }
        if (pathParameters != null) {
            payload.append(pathParameters);
        }
        if (payload.length() > 0) {
            contentType.ifPresentOrElse(
                    it -> {
                        if (it.startsWith("application/json")) {
                            request.params("body", payload.toString());
                        }
                        if (it.equals("application/x-www-form-urlencoded")) {
                            extractParameters(payload, request);
                        }
                    }, () -> System.err.println("No Content-Type found")
            );
            if (pathParameters != null) {
                extractParameters(payload, request);
            }
            if (payload.toString().contains("Content-Disposition: form-data;")) {
                System.err.println("multipart/form-data is not implemented !");
            }
        }
    }

    private static void extractParameters(StringBuilder payload, Request request) {
        final StringTokenizer pTokens = new StringTokenizer(payload.toString(), "&");
        while (pTokens.hasMoreTokens()) {
            final StringTokenizer vTokens = new StringTokenizer(pTokens.nextToken(), "=");
            if (vTokens.hasMoreTokens()) {
                final String key = vTokens.nextToken();
                if (vTokens.hasMoreTokens()) {
                    request.params(key, vTokens.nextToken());
                }
            }
        }
    }
}

File: jserver-core/src/main/kotlin/com/github/gr3gdev/jserver/http/impl/ReaderUtil.kt
package com.github.gr3gdev.jserver.http.impl

import com.github.gr3gdev.jserver.http.Request
import com.github.gr3gdev.jserver.logger.Logger
import java.io.BufferedReader
import java.io.IOException
import java.util.*

/**
 * @author Gregory Tardivel
 */
internal object ReaderUtil {

    @Throws(IOException::class)
    fun loadHeaders(request: Request, pReader: BufferedReader?) {
        var headerLine: String? = ""
        while (pReader?.readLine().also { headerLine = it }?.isNotBlank() == true) {
            val hTokens = StringTokenizer(headerLine, ":")
            var key: String
            if (hTokens.hasMoreTokens()) {
                key = hTokens.nextToken()
                if (hTokens.hasMoreTokens()) {
                    request.headers(key, hTokens.nextToken().trim())
                }
            }
        }
    }

    @Throws(IOException::class)
    fun loadParameters(request: Request, pathParameters: String?, pReader: BufferedReader?) {
        val contentType = request.headers("Content-Type")
        val payload = StringBuilder()
        while (pReader?.ready() == true) {
            payload.append(pReader.read().toChar())
        }
        if (pathParameters != null) {
            payload.append(pathParameters)
        }
        if (payload.isNotEmpty()) {
            if (contentType == null) {
                Logger.warn("No Content-Type found")
            }
            if (payload.toString().contains("Content-Disposition: form-data;")) {
                Logger.error("multipart/form-data is not implemented !")
            }
            if (contentType == "application/json") {
                request.params("body", payload.toString())
            }
            if (contentType == "application/x-www-form-urlencoded" || pathParameters != null) {
                val pTokens = StringTokenizer(payload.toString(), "&")
                while (pTokens.hasMoreTokens()) {
                    val vTokens = StringTokenizer(
                            pTokens.nextToken(), "=")
                    var key: String
                    if (vTokens.hasMoreTokens()) {
                        key = vTokens.nextToken()
                        if (vTokens.hasMoreTokens()) {
                            request.params(key, vTokens.nextToken())
                        }
                    }
                }
            }
        }
    }
}